### PR TITLE
initial diagrammer refactor

### PR DESCRIPTION
**Pull request by noahamac**
As part of general refactoring, this PR modularizes, refactors, and documents what was `diagrammer.ts`. This file has been broken up into `LookmlDiagrammer/` and better separates the functionality.

diagram_generation tests passing:
image

https://looker.atlassian.net/browse/EMBED-1614

### CODE REVIEW

File: src/utils/LookmlDiagrammer/diagrammer.ts [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+  /**
+   * A recursive function for arranging diagram tables. Starting at the base table,
+   * arrange each of its each joined tables to the left and right, switching sides for each
+   * table and leaving a gap between tables of same degree. As a table is arranged, join any of its 
+   * tables in the direction it was placed off of the base table. 
+   * @param table - view name
+   * @param degree - number of joins away from base
+   */
+  function arrangeTables(table: string, degree: number) {
```
Comment (noahamac) [inline src/utils/LookmlDiagrammer/diagrammer.ts:194]: Having difficulty figuring out how to extract this function bc it is so entangled. 

It's also the heart of the application

Perhaps it should stay? Please advise

Comment (bryans99) [reply src/utils/LookmlDiagrammer/diagrammer.ts:194]: Oh boy, thats a doozy of a function :D.

1. Consider converting arrangeTables to a class (TableArranger?). scaffold, built, shift and yOrder become properties of the class.
2. Consider passing in scaffold, built, shift and yOrder as arguments to the function arrangeTables and pass them down recursively.

Comment (noahamac) [reply src/utils/LookmlDiagrammer/diagrammer.ts]: Wow, I really like the class based implementation. TBH it's my first in JS :smiley_cat: 

Over time I will move more functionality into this class, thus more generic naming.

Comment (bryans99) [reply src/utils/LookmlDiagrammer/diagrammer.ts]: Before you add more functionality to the class consider the single responsibility principle. You may be better off have more classes. Remember that you can have compositions of classes (as an aside, composition better than inheritance, at least in my opinion).

File: src/utils/LookmlDiagrammer/minimapper.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  
+    let len = currentDegrees && currentDegrees.length
+  
+    let mid = currentDegrees && Math.ceil(len / 2);
+    
+    // figure out which degree is the middle
+    median = ((len % 2) == 0) ? (currentDegrees[mid] + currentDegrees[mid - 1]) / 2 : currentDegrees[mid - 1];
+
+    Object.keys(currentDimensions.yOrderLookup).forEach((d: string) => {
+      let degreeTablesLength = currentDimensions.yOrderLookup[d].map((tableName: string) => {
```
Comment (bryans99) [inline src/utils/LookmlDiagrammer/minimapper.ts:56]: const?

File: src/utils/LookmlDiagrammer/utils.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+/**
+ * a function which takes in explore parameters and returns a list of unique view names. 
+ * @param exploreFields - the explore field set
+ * @param joins - the explore join set
+ * @param exploreName - the name of the current explore
+ */
+export function getViews(exploreFields: ILookmlModelExploreFieldset, joins: ILookmlModelExploreJoins[], exploreName?: string) {
+  let fields = getFields(exploreFields)
```
Comment (bryans99) [inline src/utils/LookmlDiagrammer/utils.ts:64]: const?

**Review by bryans99 — CHANGES_REQUESTED** [review]
Nothing major but I have not tested it yet.

1. License headers missing
2. Use of let when const could be used

Probably want to add linter but separate PR (or adjust the rules if you already have one)

**Comment by noahamac** [comment]
Thanks for all the feedback @bryans99 . Will make a ticket for Prettier and do that on its own.... Will be a Mega-PR... Even for me :smile:

**Comment by noahamac** [comment]
Oh yeah and if you `yarn start` the extension you can use it on DemoExpo for a great diversity of models & explores:
https://demoexpo.looker.com/extensions/lookml_model_diagram::dev_ext/models/retail_banking/explores/card_order_sequence

File: src/utils/LookmlDiagrammer/diagrammer.ts (hunk 1 above)
Comment (noahamac) [reply src/utils/LookmlDiagrammer/diagrammer.ts]: Yes, I would want LookmlDiagrammer as orchestration class that invokes JoinUntangler TableArranger ViewBuilder etc

Comment (noahamac) [reply src/utils/LookmlDiagrammer/diagrammer.ts]: Future work for now tho. More diagrammer classes would help in implementing Edit mode (v2)

**Review by bryans99 — APPROVED** [review]
LGTM

**Pull request merged**